File: 05_April_1st/gyeongwook/insertionSortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        part1 = []
        part2 = []
        
        currentNode = head
        returnValue = head

        while currentNode is not None :
            nextNode = currentNode.next

            if len(part2) == 0 or currentNode.val <= part1[-1].val :
                while len(part1) is not 0 and currentNode.val < part1[-1].val :
                    part2.append(part1.pop())
            else :
                while len(part2) is not 0 and currentNode.val > part2[-1].val :
                    part1.append(part2.pop())
            
            if len(part1) is not 0 :
                part1[-1].next = currentNode
            else :
                returnValue = currentNode
            
            part1.append(currentNode);
            
            if len(part2) is not 0 :
                currentNode.next = part2[-1]
            else :
                currentNode.next = None
                
            currentNode = nextNode;
            
        return returnValue;

File: 05_April_1st/gyeongwook/insertionSortList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *insertionSortList(ListNode *head) {
        ListNode* sentinel = new ListNode(0);
        ListNode* current = head;
        ListNode* next = nullptr;
        ListNode* prev = nullptr;
        
        while (current != nullptr) {
            next = current->next;
            prev = sentinel;
            
            while (prev->next != nullptr && prev->next->val <= current->val)
                prev = prev->next;
            
            current->next = prev->next;
            prev->next = current;
            current = next;
        }
        
        ListNode* ret = sentinel->next;
        delete sentinel;
        
        return ret;
    }
};